query-engine/query-engine-c-abi/src/engine.rs

```diff
@@ -0,0 +1,674 @@
+}
+
+fn get_cstr_safe(ptr: *const c_char) -> Option<String> {
+    if ptr.is_null() {
+        None
+    } else {
+        let cstr = unsafe { CStr::from_ptr(ptr) };
+        Some(String::from_utf8_lossy(cstr.to_bytes()).to_string())
+    }
+}
+
+fn map_known_error(err: query_core::CoreError) -> crate::Result<String> {
+    let user_error: user_facing_errors::Error = err.into();
+    let value = serde_json::to_string(&user_error)?;
+
+    Ok(value)
+}
+
+// Struct that holds an internal prisma engine
+// the inner prop holds the internal state, it starts as a Builder
+// meaning it is not connected to the database
+// a call to connect is necessary to start executing queries
+impl QueryEngine {
+    /// Parse a valid datamodel and configuration to allow connecting later on.
+    pub fn new(constructor_options: ConstructorOptions) -> Result<Self, Error> {
+        // Create a logs closure that can be passed around and called at any time
+        // safe scheduling should be taken care by the code/language/environment calling this C-compatible API
+        let engine_id = get_cstr_safe(constructor_options.id).expect("engine id cannot be missing");
+        let log_callback_c = constructor_options.log_callback;
+        let log_callback = move |msg: String| {
+            let id = CString::new(engine_id.clone()).unwrap();
+            let c_message = CString::new(msg).unwrap();
+            unsafe {
+                log_callback_c(id.as_ptr(), c_message.as_ptr());
+            }
+        };
+
+        let str_env = get_cstr_safe(constructor_options.env).expect("Environment missing");
+        let json_env = serde_json::from_str(str_env.as_str()).expect("Environment cannot be parsed");
+        let env = stringify_env_values(json_env)?; // we cannot trust anything JS sends us from process.env
+
+        let str_datasource_overrides =
+            get_cstr_safe(constructor_options.datasource_overrides).expect("Datesource overrides missing");
+        let json_datasource_overrides =
+            serde_json::from_str(str_datasource_overrides.as_str()).expect("Datasource overrides cannot be parsed");
+        let overrides: Vec<(_, _)> = stringify_env_values(json_datasource_overrides)
+            .unwrap()
+            .into_iter()
+            .collect();
+
+        let datamodel = get_cstr_safe(constructor_options.datamodel).expect("Datamodel must be present");
+        let mut schema = psl::validate(datamodel.into());
+        let config = &mut schema.configuration;
+
+        // let preview_features = config.preview_features();
+
+        schema
+            .diagnostics
+            .to_result()
+            .map_err(|err| ApiError::conversion(err, schema.db.source()))?;
+
+        let base_path = get_cstr_safe(constructor_options.base_path);
+        match &base_path {
+            Some(path) => env::set_current_dir(Path::new(&path)).expect("Could not change directory"),
+            _ => println!("No base path passed!"),
```

Comment:
Not really sure how to call the logger on this level, or do you mean just call the passed log function on this level. If you could help me with a snippet or an example it would be great.

Reply:
I don't think this is something that we want to forward to the PrismaClient. Therefore, `tracing::trace!("No base path passed")` should be enough.